### PR TITLE
KXI-69709 PNGs from matplotlib can now be displayed in a kxnb notebook

### PR DESCRIPTION
**Pull request by bjeffery-kx**
### Changes introduced by this PR
- Regardless of the returnFormat, when a Python call returns a byte array starting with a PNG signature, that PNG will be returned serialized and will be rendered in notebooks
- evaluatePy.q has been split up so there's one function to evaluate code, and another to format it. This makes it much simpler to change the return format based on the result type.
- This also appears (unintentionally) to fix a bug where stack traces weren't being shown for errors in Python notebook cells
- The behaviour of evaluatePy.q should be otherwise unchanged

This does mean evaluate.q diverges more from framework. I couldn't find the code to build evaluatePy.q from what's in framework. Ideally, the same refactor could be applied to framework, since it makes the code simpler to work with.

# Testing
I've tested this on Windows, and from a Python notebook cell I was able to display any PNG with
```
kx.q("read1 `:/path/to/image.png")
```
or this in a q cell
```
read1 `:/path/to/image.png
```

From a .q file, ``read1 `:/path/to/image.png`` will print the stringified byte string to the console, and if KDB Results is selected, it will open the image in a tab named plot-1.plot

From a .py file with the terminal focused however, displaying a png, such as with ``kx.q("read1 `:/path/to/image.png")`` will fail with the error "TypeError: t.replace is not a function". It displays the image correctly if the KDB RESULTS tab is focused.

As the refactor touched a lot of evaluatePy.q, I've tested this with a variety of long and complex Python expressions as well as edge cases, and all are working except the above PNG case from .py files

### CODE REVIEW

File: test/q/tests/evaluatePy.quke [hunk 1]
```diff
@@ -597,9 +582,11 @@ feature python toStructuredText
                 "{\"count\":13,\"columns\":[{\"name\":\"values\",\"type\":\"str\",\"values\":[\"'The brown fox'\"],\"order\":[0]}],\"warnings\":[]}"]
         expect python structured text to truncate all other data types when count of data is over TABULAR_LIMIT
             .struct.TABULAR_LIMIT : 2;
-            .qu.compare[
+            result: .qu.compare[
                 .test.evaluatePy["structuredText"; "(1,2,3,4,5)"]`data;
-                "{\"count\":5,\"columns\":[{\"name\":\"values\",\"type\":\"tuple\",\"values\":[\"1\",\"2\"],\"order\":[0,1]}],\"warnings\":[\"Results truncated to TABULAR_LIMIT. Console view is faster for large data.\"]}"]
+                "{\"count\":5,\"columns\":[{\"name\":\"values\",\"type\":\"tuple\",\"values\":[\"1\",\"2\"],\"order\":[0,1]}],\"warnings\":[\"Results truncated to TABULAR_LIMIT. Console view is faster for large data.\"]}"];
+            .struct.TABULAR_LIMIT : 600000;
```
Comment (acheung12) [inline test/q/tests/evaluatePy.quke:588]: why did we need to increase the limit here? What does this change solve?

File: package.json [hunk 2]
```diff
@@ -1315,4 +1315,4 @@
     "undici": "7.28.0",
     "form-data": ">=4.0.6"
   }
-}
```
Comment (acheung12) [inline package.json:1318]: revert change?

File: CHANGELOG.md [hunk 3]
```diff
@@ -34,6 +34,12 @@ file.
   `\someCommand "\t"`) was not escaped correctly when the command was rewritten
   for execution.
 
+# v1.18.1
+
+### Enhancements
+
+- VS Code endpoint detection now proactively checks for `.vscode.getManifest` availability rather than falling back to `vscode.q` only after `123` execution fails.
+
```
Comment (acheung12) [inline CHANGELOG.md:42]: Is this part of the changes to this ticket?

File: test/q/tests/evaluatePy.quke (hunk 1 above)
Comment (bjeffery-kx) [reply test/q/tests/evaluatePy.quke:588]: This is to undo the `.struct.TABULAR_LIMIT` at the start of the test, otherwise the new structuredText test gets truncated at 2 rows. I considered other ways to do this, this could be put into an `after each` block, but then you can't step through the tests as easily.

I also considered adding this to the `after` block, then putting the new tests in their own feature, but thought it would be a good idea generally to have this test reset `struct.TABULAR_LIMIT` in case new tests were added to this feature.

File: resources/q/evaluatePy.q [hunk 4]
```diff
@@ -1,129 +1,137 @@
...
+        return code
+
+    def pystruct_format_exception():
+        type, error, tb = sys.exc_info()
+        stacktrace = traceback.extract_tb(tb)
+        offset = -1 * (len(stacktrace) - 3)
+        formatted_tb = traceback.format_exception(type, error, tb, offset)
+        formatted_tb = ''.join(formatted_tb)
+        # encode('UTF-8') is to return a string, to avoid polluting the symbol table.
+        return formatted_tb.rstrip().encode('UTF-8')
```
Comment (acheung12) [inline resources/q/evaluatePy.q:49]: Good idea.

File: resources/q/evaluatePy.q [hunk 5]
```diff
@@ -327,102 +333,126 @@
...
+                (`data;::);
+                (`stacktrace;.pykx.qeval["lambda r: r['stacktrace']"]result));
+            (!). flip(
+                (`error;0b);
+                (`errorMsg;errorMsg);
+                (`data;.pykx.eval["lambda r: r['data']"]result))]
+        };
+
+    // text and serialized results return the stacktrace as a symbol, structuredText as a string
+    format:{[returnFormat; sampleFn; sampleSize; result]
```
Comment (acheung12) [inline resources/q/evaluatePy.q:407]: sampleFn is different from the current formatting style we have; which is sample_fn. Same for sampleSize

File: resources/q/evaluatePy.q [hunk 6]
```diff
@@ -327,102 +333,126 @@
...
+                    text],"\n"
+                }(),.pykx.qeval["_kx_execution_context['pystruct_to_text']"]result`data;
+            :result];
+        
+        if [returnFormat ~ "serialized";
+            result[`data]: result[`data]`;
+            :result];
+    
+        if [returnFormat ~ "structuredText";
+            result:.pykx.qeval["pystruct_format_structured_text"][result`data;sampleFn;sampleSize];
```
Comment (acheung12) [inline resources/q/evaluatePy.q:424]: sampleFn is different from the current formatting style we have; which is sample_fn. Same for sampleSize

File: resources/q/evaluatePy.q [hunk 7]
```diff
@@ -327,102 +333,126 @@
...
+
+    // This is a hack to display PNGs, such as those generated by matplotlib
+    // regardless of what the expected return type is.
+    // It should be replaced by PNGs coming back over websocket.
+    if [.pykx.qeval["is_png_bytevector"]result `data;
+        returnFormat: "serialized"];
+    
+    response: $[result`error;
+        result;
+        format[returnFormat; sampleFn; sampleSize; result]];
```
Comment (acheung12) [inline resources/q/evaluatePy.q]: sampleFn is different from the current formatting style we have; which is sample_fn. Same for sampleSize

File: resources/q/evaluatePy.q [hunk 8]
```diff
@@ -327,102 +333,126 @@
...
+        };
+    
+    code:removeExtraIndents code;
+    defaultConv:.pykx.util.defaultConv;
+    .pykx.util.defaultConv:"k";
+    result: evaluate code;
+    
+    // This can't be replaced with something simpler, like
+    // (type[result[`data]`]=4h) and 0x89504e470d0a1a0a ~ 8#result[`data]`;
+    // because there are side-effects and overhead to casting a some values to q, and some values will error when cast to q
```
Comment (acheung12) [inline resources/q/evaluatePy.q]:
```suggestion
    // because there are side-effects and overhead to casting some values to q, and some values will error when cast to q
```

File: resources/q/evaluatePy.q [hunk 9]
```diff
@@ -327,102 +333,126 @@
...
+                (`error;0b);
+                (`errorMsg;errorMsg);
+                (`data;.pykx.eval["lambda r: r['data']"]result))]
+        };
+
+    // text and serialized results return the stacktrace as a symbol, structuredText as a string
+    format:{[returnFormat; sampleFn; sampleSize; result]
+        if [returnFormat ~ "text";
+            result[`data]: {[text]
+                $[  count[text]>maxSize:250000;
```
Comment (acheung12) [inline resources/q/evaluatePy.q]: What's this limit? Need an explanation

File: resources/q/evaluatePy.q [hunk 10]
```diff
@@ -327,102 +333,126 @@
...
+    // regardless of what the expected return type is.
+    // It should be replaced by PNGs coming back over websocket.
+    if [.pykx.qeval["is_png_bytevector"]result `data;
+        returnFormat: "serialized"];
+    
+    response: $[result`error;
+        result;
+        format[returnFormat; sampleFn; sampleSize; result]];
+    
+    .pykx.util.defaultConv:defaultConv;
```
Comment (acheung12) [inline resources/q/evaluatePy.q:469]: Why is this set at the very end?

**Comment by ecmel** [comment]
@bjeffery-kx Used the Python wrapper on IE scratchpad but the limit still hits and PNGs truncates, so increased 10K to 100K.

File: package.json (hunk 2 above)
Comment (bjeffery-kx) [reply package.json:1318]: I'm not sure if this is github saying the file formerly didn't end in a newline, or if there was an extra newline removed. Either way, this version correctly ends in a newline.

File: CHANGELOG.md (hunk 3 above)
Comment (bjeffery-kx) [reply CHANGELOG.md:42]: No, I'm not sure how it ended up in this MR, but it should be added to CHANGELOG.md

File: resources/q/evaluatePy.q (hunk 5 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q:407]: I've updated the q variables to use the q style. The Python variables still use the Python style.

File: resources/q/evaluatePy.q (hunk 6 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q:424]: I've updated the q variables to use the q style. The Python variables still use the Python style.

File: resources/q/evaluatePy.q (hunk 7 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q]: I've updated the q variables to use the q style. The Python variables still use the Python style.

File: resources/q/evaluatePy.q (hunk 10 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q:469]: In the stream processor, where we create the process and can make changes to it freely, we set .pykx.util.defaultConv to "k", and leave it that way.

With VS Code, since we're connecting to processes we don't own, we try not to make any changes to it. The original value gets cached before we run the user's code
```
    defaultConv:.pykx.util.defaultConv;
    .pykx.util.defaultConv:"k";
```
Then we restore it afterwards

**Review by ecmel — DISMISSED** [review]
We should merge https://github.com/KxSystems/kx-vscode/pull/779 before this one.

File: resources/q/evaluatePy.q (hunk 9 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q]: I've added a comment explaining it

**Review by ecmel — DISMISSED** [review]

File: src/classes/localConnection.ts [hunk 11]
```diff
@@ -192,6 +192,11 @@ export class LocalConnection {
           const result = res.data === null ? "" : res.data;
           if (stringify) {
             resolve(result);
+          } else if (Array.isArray(result)) {
+            // This handles the case where the response is an array of numbers.
+            // The back end will disregard the return format and return an array of bytes
+            // if the result of the expression is a byte array starting with the PNG signature 0x89504e470d0a1a0a
+            resolve(result);
```
Comment (acheung12) [inline src/classes/localConnection.ts]: why can't we do something like this? 

if (stringify || Array.isArray(result))

File: resources/q/evaluatePy.q (hunk 10 above)
Comment (acheung12) [reply resources/q/evaluatePy.q:469]: Why don't we trap it? What happens if something breaks in between? call .Q.trp?

If anything between the set and the restore signals a q error, defaultConv stays "k" in a process you don't own

File: resources/q/evaluatePy.q [hunk 12]
```diff
@@ -327,102 +333,128 @@
...
+            :result];
+        
+        // Unrecognized return format
+        : `error`errorMsg`data!(1b; "Invalid returnFormat specified"; ::)
+        };
+    
+    code:removeExtraIndents code;
+    defaultConv:.pykx.util.defaultConv;
+    .pykx.util.defaultConv:"k";
+    result: evaluate code;
```
Comment (acheung12) [inline resources/q/evaluatePy.q]: We evaluate the user code and then determine if the returnFormat is valid or invalid. This is going to cause side effects if we reject a bad returnformat. Usercode will still be ran and could have side effects such as mutated globals or other things.

File: resources/q/evaluatePy.q [hunk 13]
```diff
@@ -327,102 +333,128 @@
...
+    // This can't be replaced with something simpler, like
+    // (type[result[`data]`]=4h) and 0x89504e470d0a1a0a ~ 8#result[`data]`;
+    // because there are side-effects and overhead to casting some values to q, and some values will error when cast to q
+    .pykx.pyexec "def is_png_bytevector(x=None):
+        import pykx as kx
+
+        return (
+            isinstance(x, kx.ByteVector)
+            and len(x) >= 8
+            and bytes(x.py()[:8]) == bytes.fromhex('89504e470d0a1a0a')
```
Comment (acheung12) [inline resources/q/evaluatePy.q]: we can simplify this by just converting the 8 bytes rather than converting the entire vector. 

x[:8].py()

File: resources/q/evaluatePy.q [hunk 14]
```diff
@@ -327,102 +333,128 @@
...
+    
+    code:removeExtraIndents code;
+    defaultConv:.pykx.util.defaultConv;
+    .pykx.util.defaultConv:"k";
+    result: evaluate code;
+    
+    // This can't be replaced with something simpler, like
+    // (type[result[`data]`]=4h) and 0x89504e470d0a1a0a ~ 8#result[`data]`;
+    // because there are side-effects and overhead to casting some values to q, and some values will error when cast to q
+    .pykx.pyexec "def is_png_bytevector(x=None):
```
Comment (acheung12) [inline resources/q/evaluatePy.q:436]: why is x=None present? Can't we achieve the same thing by guarding the call with

`not result`error``

**Review by acheung12 — CHANGES_REQUESTED** [review]

File: test/q/tests/evaluatePy.quke [hunk 15]
```diff
@@ -628,11 +615,22 @@ feature python toStructuredText
...
-            
\ No newline at end of file
+
+    should handle PNGs returned regardless of the returnFormat
+        expect byte strings which aren't PNGs to be returned as structuredText
+            .qu.compare[
+                .test.evaluatePy["structuredText"; "kx.q('0x89504e470d0a1a')"][`data];
+                "{\"count\":7,\"columns\":[{\"name\":\"values\",\"type\":\"ByteVector\",\"values\":[\"pykx.ByteAtom(pykx.q('0x89'))\",\"pykx.ByteAtom(pykx.q('0x50'))\",\"pykx.ByteAtom(pykx.q('0x4e'))\",\"pykx.ByteAtom(pykx.q('0x47'))\",\"pykx.ByteAtom(pykx.q('0x0d'))\",\"pykx.ByteAtom(pykx.q('0x0a'))\",\"pykx.ByteAtom(pykx.q('0x1a'))\"],\"order\":[5,4,6,3,2,1,0]}],\"warnings\":[]}"]
+        expect PNG byte strings to be returned serialized, regardless of the return format
+            all (
```
Comment (acheung12) [inline test/q/tests/evaluatePy.quke]: can we split this into 3 expect blocks so we know which specific error occurs?

File: resources/q/evaluatePy.q [hunk 16]
```diff
@@ -327,102 +333,128 @@
...
+                (`errorMsg;errorMsg);
+                (`data;.pykx.eval["lambda r: r['data']"]result))]
+        };
+
+    // text and serialized results return the stacktrace as a symbol, structuredText as a string
+    format:{[returnFormat; sampleFn; sampleSize; result]
+        if [returnFormat ~ "text";
+            result[`data]: {[text]
+                // If the text is longer than 250,000 characters,
+                // truncate the text and an a q style ".."
```
Comment (acheung12) [inline resources/q/evaluatePy.q]: typo. Is it suppose to be "and add a q style"?

File: resources/q/evaluatePy.q [hunk 17]
```diff
@@ -327,102 +333,128 @@
...
+                $[  count[text]>maxSize:250000;
```
Comment (acheung12) [inline resources/q/evaluatePy.q]: Assigning it inline is strange and hard to read. we should leave it as a constant higher up like we did before.

File: src/classes/localConnection.ts (hunk 11 above)
Comment (bjeffery-kx) [reply src/classes/localConnection.ts]: We definitely could, and that's how I wrote it originally, but then I separated them so it was clearer what the comment referred to. I'll switch it back

**Review by acheung12 — DISMISSED** [review]

File: resources/q/evaluatePy.q (hunk 12 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q]: granted, but the alternatives get kind of messy. We're then listing the return formats in multiple places, and potentially adding a early return or at least extra conditionals to skip over evaluation and formatting.

It also means the error checking for format is separated from the format function. Not necessarily a bad thing, we do that other places, but it's trading one issue for another.

returnFormat isn't ever set by the user, so it's more a case of "this line will execute if something goes disastrously wrong, and we'll get a clear error message in the MR tests", instead of a "confirm the user entered a valid email address" error that's expected to happen routinely.

File: resources/q/evaluatePy.q (hunk 14 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q:436]: Without `x=None`, calling this function with generic null throws the error
```
TypeError("is_png_bytevector() missing 1 required positional argument: 'x'")
  [1]  .pykx.qeval["is_png_bytevector"][::]
       ^
```

This could be replaced with a guard like
```
if [not result[`data] ~ (::);
    if [.pykx.qeval["is_png_bytevector"]result `data;
        returnFormat: "serialized"]];
```
but then the caller of the function needs to know how to call it safely, and if it were called from multiple places, each caller would need to include the guard. It's easier to have a function that can safely accept any q type.

Comment (bjeffery-kx) [reply resources/q/evaluatePy.q:436]: I'll add a comment about why there's a default parameter

File: resources/q/evaluatePy.q (hunk 13 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q]: Interestingly, there's no effect on performance (runtime or memory usage). I think it's because the conversion is just done on the vector's metadata, and the underlying bytes are unchanged. Likewise, I think taking the first eight bytes does a lazy copy where it creates a new header pointed at the same underlying data.

I'll make the change though, as future maintainers are likely to have the same thought.

File: test/q/tests/evaluatePy.quke (hunk 15 above)
Comment (bjeffery-kx) [reply test/q/tests/evaluatePy.quke]: done

File: resources/q/evaluatePy.q (hunk 16 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q]: fixed

File: resources/q/evaluatePy.q (hunk 17 above)
Comment (bjeffery-kx) [reply resources/q/evaluatePy.q]: fixed

File: resources/q/evaluatePy.q (hunk 14 above)
Comment (acheung12) [reply resources/q/evaluatePy.q:436]: Sounds good. Yeah, if we were only calling in this in one spot, i would be more inclined to use a guard,  but if we potentially might need to call this in multiple places, a default value could be useful. Either way a comment explaining this is sufficient.

**Comment by acheung12** [comment]
All threads resolved. I'm fine with merging with the catch that the typescript error gets resolved. The focusing of the .py tab while running code should be fixed.

**Review by ecmel — APPROVED** [review]
